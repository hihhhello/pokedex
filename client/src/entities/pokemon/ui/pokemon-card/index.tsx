import {
  CardHeader,
  SxProps,
  Card,
  CardMedia,
  CardActions,
} from '@mui/material';
// @ts-ignore
import defaultPokemonPng from 'assets/default-pokemon.png';
import React from 'react';

export interface Props {
  name: string;
  avatarUrl?: string | null;
  actions?: React.ReactNode[];
  sx?: SxProps;
}

export const PokemonCard = ({ actions = [], name, avatarUrl, sx }: Props) => {
  return (
    <Card sx={sx}>
      <CardHeader title={name} />
      <CardMedia component="img" src={avatarUrl || defaultPokemonPng} />
      <CardActions>
        {actions.map((item, i) => (
          <React.Fragment key={i}>{item}</React.Fragment>
        ))}
      </CardActions>
    </Card>
  );
};
