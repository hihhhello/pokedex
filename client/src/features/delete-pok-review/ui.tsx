import { Delete } from '@mui/icons-material';
import { IconButton } from '@mui/material';
import { pokemonModel } from 'entities/pokemon';
import { toast } from 'react-toastify';

interface Props {
  reviewId: number;
  fetchRatings: (pokemonId: number) => Promise<void>;
  fetchPokemon: (apiId: number) => Promise<void>;
}

export const DeletePokReview = (props: Props) => {
  const { fetchPokemon, fetchRatings, reviewId } = props;

  const deleteReview = async () => {
    const data = await pokemonModel.deletePokemonRating(reviewId);
    if (data) {
      await fetchPokemon(data.favPokemon.apiId);
      await fetchRatings(data.favPokemon.id);
      toast('Review was deleted', {
        type: 'success',
        toastId: 'delete-pok-review',
      });
    }
  };

  return (
    <IconButton onClick={deleteReview}>
      <Delete />
    </IconButton>
  );
};
