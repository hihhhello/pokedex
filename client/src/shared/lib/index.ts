export * from './testing-utils';
export * from './constants';
export * from './utils';
export * from './hooks';
