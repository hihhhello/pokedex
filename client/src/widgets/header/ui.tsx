import React from 'react';
import { Logout } from '@mui/icons-material';
import { Stack, IconButton, Typography } from '@mui/material';
import { userModel } from 'entities/user';
import { NavLink } from 'react-router-dom';

export const Header = () => {
  const { logout, user, authorizing } = userModel.context.useUserContext();

  const onLogout = () => {
    logout();
  };

  return (
    user && (
      <Stack
        direction="row"
        alignItems="center"
        justifyContent="space-between"
        sx={{
          height: '10vh',
          backgroundColor: 'primary.dark',
          paddingBlock: 1,
          paddingInline: 2,
        }}
      >
        {!authorizing && (
          <>
            <Typography
              sx={{
                a: {
                  color: 'common.white',
                  ':hover': {
                    color: 'grey.500',
                  },
                  transition: '.3s all',
                },
              }}
            >
              <NavLink to="/">Home</NavLink>
            </Typography>
            <Typography
              sx={{
                a: {
                  color: 'common.white',
                  flex: 1,
                  textTransform: 'uppercase',
                  textDecoration: 'none',
                  transition: '.3s all',
                  ':hover': {
                    color: 'grey.500',
                  },
                  fontSize: 18,
                },
              }}
            >
              <NavLink to={`/${user!.login}`} data-testid="header-login">
                {user!.login}
              </NavLink>
            </Typography>
            <IconButton
              sx={{ color: 'common.white' }}
              size="large"
              data-testid="header-logout"
              onClick={onLogout}
            >
              <Logout />
            </IconButton>
          </>
        )}
      </Stack>
    )
  );
};
