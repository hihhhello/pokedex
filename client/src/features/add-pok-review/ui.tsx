import { Rating, TextField, Box, Button, SxProps } from '@mui/material';
import { pokemonModel } from 'entities/pokemon';
import React, { useState } from 'react';
import { toast } from 'react-toastify';
import { AddFavPokemonDto } from 'shared/api';

// TODO: refactor add --> fetch flow with redux or another context if needed
// For a showcase fetch functions in props is enough

interface Props {
  userId: number;
  pokemon: AddFavPokemonDto;
  fetchRatings: (pokemonId: number) => Promise<void>;
  fetchPokemon: (apiId: number) => Promise<void>;
  sx?: SxProps;
}

export const AddPokReview = (props: Props) => {
  const { fetchRatings, pokemon, userId, fetchPokemon } = props;

  const [loading, setLoading] = useState(false);
  const [text, setText] = useState('');
  const [rating, setRating] = useState<number | null>(null);
  const [error, setError] = useState<null | string>(null);

  const addPokemon = async () => {
    if (!rating) {
      setError('Please, select rating');
      toast('You have issue in your review', {
        type: 'error',
        toastId: 'input-error',
      });
      return;
    }

    setLoading(true);
    const data = await pokemonModel.addPokemonRating({
      rating,
      pokemon,
      text,
      user: {
        id: userId,
      },
    });

    if (data) {
      await fetchRatings(data.favPokemon.id);
      await fetchPokemon(data.favPokemon.apiId);
      toast('New review was added!', {
        type: 'success',
        toastId: 'added-new-review',
      });
    }

    setLoading(false);
  };

  const onRatingChange = (_: any, newValue: number | null) => {
    setRating(newValue);
  };

  const onTextChange = (e: React.ChangeEvent<HTMLInputElement>) => {
    setText(e.target.value);
  };

  return (
    <Box
      sx={{
        display: 'flex',
        flexDirection: 'column-reverse',
        alignItems: 'center',
        justifyContent: 'center',
        width: '100%',
        gap: 2,
        ...props.sx,
      }}
    >
      <Box
        sx={{
          display: 'flex',
          alignItems: 'center',
          justifyContent: 'center',
          gap: 2,
          width: '100%',
        }}
      >
        <TextField
          helperText={error}
          error={Boolean(error)}
          value={text}
          onChange={onTextChange}
          size="small"
        />
        <Button
          onClick={addPokemon}
          disabled={loading}
          variant="contained"
          size="medium"
        >
          {loading ? 'Loading...' : 'Add'}
        </Button>
      </Box>

      <Rating value={rating} onChange={onRatingChange} size="large" />
    </Box>
  );
};
