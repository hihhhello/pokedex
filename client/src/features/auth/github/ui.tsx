import React from 'react';
import { GitHub } from '@mui/icons-material';
import { Button } from '@mui/material';
import { GITHUB_OAUTH_LINK } from 'shared/lib';

export const GithubAuthButton = () => {
  const onClickHandler = () => {
    window.location.replace(GITHUB_OAUTH_LINK);
  };

  return (
    <Button
      data-testid="github-auth-btn"
      variant="contained"
      startIcon={<GitHub />}
      onClick={onClickHandler}
    >
      Sign In via GitHub OAuth
    </Button>
  );
};
