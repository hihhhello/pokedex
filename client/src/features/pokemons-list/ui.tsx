import React, { useCallback, useEffect, useRef, useState } from 'react';
import { PokemonClient, NamedAPIResource } from 'pokenode-ts';
import { useNavigate } from 'react-router-dom';
import { columns } from './lib';
import {
  Paper,
  Table,
  TableBody,
  TableCell,
  TableContainer,
  TableHead,
  TablePagination,
  TableRow,
  SxProps,
  CircularProgress,
  Stack,
  Box,
  IconButton,
} from '@mui/material';
import { Visibility } from '@mui/icons-material';
import { toast } from 'react-toastify';
import { useDebounce } from 'shared/lib';
import { pokemonModel, pokemonLib } from 'entities/pokemon';

interface Props {
  PaperSx?: SxProps;
  TableContainerSx?: SxProps;
  PaginationSx?: SxProps;
}

export const List = ({ PaperSx, TableContainerSx, PaginationSx }: Props) => {
  const {
    page,
    rowsPerPage,
    totalCount,
    setTotalCount,
    setPage,
    setRowsPerPage,
  } = pokemonModel.usePokemonContext();

  const navigate = useNavigate();

  const pokemonClient = new PokemonClient();

  const [loading, setLoading] = useState(false);

  const [pokemons, setPokemons] = useState<NamedAPIResource[]>([]);

  const debouncedPage = useDebounce(page, 300);
  const isMounted = useRef(false);

  const fetchPokemons = async (page: number, rowsPerPage: number) => {
    try {
      setLoading(true);
      const data = await pokemonClient.listPokemons(
        page * rowsPerPage,
        rowsPerPage
      );
      setPokemons(data.results);
      setTotalCount(data.count);
    } catch (e: any) {
      toast(e.message, {
        type: 'error',
        toastId: 'pokemons-list-error',
      });
    } finally {
      setLoading(false);
    }
  };

  useEffect(() => {
    fetchPokemons(page, rowsPerPage);
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [rowsPerPage]);

  useEffect(() => {
    if (isMounted.current) {
      fetchPokemons(debouncedPage, rowsPerPage);
      return;
    }

    isMounted.current = true;

    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [debouncedPage]);

  const onPageChangeHandler = (
    event: React.MouseEvent<HTMLButtonElement> | null,
    newPage: number
  ) => {
    setPage(newPage);
  };

  const onPageSizeChangeHandler = (
    event: React.ChangeEvent<HTMLInputElement | HTMLTextAreaElement>
  ) => {
    setRowsPerPage(parseInt(event.target.value, 10));
    setPage(0);
  };

  const getPokemonDetailViewClick = useCallback(
    (name: string) => () => {
      navigate(`/pokemon/${name}`);
    },
    // eslint-disable-next-line react-hooks/exhaustive-deps
    []
  );

  return (
    <Paper
      sx={{ overflow: 'hidden', ...PaperSx }}
      data-testid="pokemons-list-paper"
    >
      <TableContainer sx={TableContainerSx}>
        <Table stickyHeader>
          <TableHead>
            <TableRow>
              {columns.map((column) => (
                <TableCell
                  data-testid="pokemons-list-head-cell"
                  key={column.id}
                  align={column.align}
                  style={{ minWidth: column.minWidth }}
                  sx={{
                    fontWeight: 600,
                    fontSize: 18,
                  }}
                >
                  <Stack direction="row" justifyContent="space-between">
                    {column.label}
                    {loading && <CircularProgress size={20} />}
                  </Stack>
                </TableCell>
              ))}
            </TableRow>
          </TableHead>
          <TableBody>
            {pokemons.map((row) => (
              <TableRow
                data-testid="pokemons-list-body-row"
                key={row.name}
                sx={{
                  '&:last-child td, &:last-child th': { border: 0 },
                }}
              >
                <TableCell component="th" scope="row">
                  <Box
                    sx={{
                      justifyContent: 'space-between',
                    }}
                  >
                    <span data-testid="pokemons-list-body-row-span">
                      {row.name}
                    </span>
                    <IconButton
                      size="small"
                      sx={{ marginLeft: 2 }}
                      onClick={getPokemonDetailViewClick(row.name)}
                    >
                      <Visibility />
                    </IconButton>
                  </Box>
                </TableCell>
              </TableRow>
            ))}
          </TableBody>
        </Table>
      </TableContainer>
      <TablePagination
        data-testid="pokemons-list-pagination"
        component="div"
        count={totalCount}
        page={page}
        onPageChange={onPageChangeHandler}
        rowsPerPage={rowsPerPage}
        onRowsPerPageChange={onPageSizeChangeHandler}
        rowsPerPageOptions={pokemonLib.rowsPerPageOptions}
        sx={{ ...PaginationSx }}
      />
    </Paper>
  );
};
