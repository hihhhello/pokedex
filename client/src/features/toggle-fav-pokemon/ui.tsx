import React, { useEffect, useRef } from 'react';
import { Checkbox } from '@mui/material';
import { Favorite, FavoriteBorder } from '@mui/icons-material';
import { toast } from 'react-toastify';
import { githubPokedexApi, AddFavPokemonDto } from 'shared/api';
import { userModel } from 'entities/user';
import { parseAxiosError, useDebounce } from 'shared/lib';
import { AxiosError } from 'axios';
import { pokemonModel } from 'entities/pokemon';

interface Props {
  pokemonData: AddFavPokemonDto;
  defaultChecked?: boolean;
}

export const ToggleFavPokemon = React.memo(
  ({ pokemonData, defaultChecked }: Props) => {
    const { user } = userModel.context.useUserContext();
    const { fetchFavPokemonsApiIds } = pokemonModel.usePokemonContext();

    const [checked, setChecked] = React.useState(defaultChecked);

    const debouncedChecked = useDebounce(checked, 300);

    const isMounted = useRef(false);

    const onAddFav = async () => {
      try {
        await githubPokedexApi.addFavPokemon(user!.id, pokemonData);
        await fetchFavPokemonsApiIds(user!.id);
        toast(`${pokemonData.name} was added to favorites`, {
          type: 'success',
          toastId: `${pokemonData.name}-success-fav-add`,
        });
      } catch (e: any) {
        if (e instanceof AxiosError) {
          const { statusText, message } = parseAxiosError(e);

          toast(message || statusText, {
            type: 'error',
            toastId: `${pokemonData.name}-add-to-fav-error`,
          });
          return;
        }
        toast(e.message, {
          type: 'error',
          toastId: `${pokemonData.name}-add-to-fav-error`,
        });
      }
    };

    const onDeleteFav = async () => {
      try {
        await githubPokedexApi.deleteFavPokemon(user!.id, pokemonData.apiId);
        await fetchFavPokemonsApiIds(user!.id);
        toast(`${pokemonData.name} was removed from favorites`, {
          type: 'success',
          toastId: `${pokemonData.name}-success-fav-delete`,
        });
      } catch (e: any) {
        if (e instanceof AxiosError) {
          const { statusText, message } = parseAxiosError(e);

          toast(message || statusText, {
            type: 'error',
            toastId: `${pokemonData.name}-delete-fav-error`,
          });
          return;
        }
        toast(e.message, {
          type: 'error',
          toastId: `${pokemonData.name}-delete-fav-error`,
        });
      }
    };

    useEffect(() => {
      if (isMounted.current) {
        if (debouncedChecked) {
          onAddFav();
        } else {
          onDeleteFav();
        }
        return;
      }

      isMounted.current = true;
      // eslint-disable-next-line react-hooks/exhaustive-deps
    }, [debouncedChecked]);

    const onCheckboxChange = (event: React.ChangeEvent<HTMLInputElement>) => {
      setChecked(event.target.checked);
    };

    return (
      <Checkbox
        data-testid="toggle-fav-pokemon"
        icon={<FavoriteBorder />}
        checkedIcon={<Favorite />}
        checked={checked}
        // @ts-ignore
        onClick={onCheckboxChange}
      />
    );
  }
);
