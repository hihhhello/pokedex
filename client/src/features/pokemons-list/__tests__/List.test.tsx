import { screen } from '@testing-library/react';
import { customRender } from 'shared/lib';
import { List } from '../ui';

describe('<PokemonsList />', () => {
  test('Renders', () => {
    customRender({
      node: <List />,
    });

    const element = screen.getByTestId('pokemons-list-paper');

    expect(element).toBeInTheDocument();
    expect(element).toMatchSnapshot();
  });
});
