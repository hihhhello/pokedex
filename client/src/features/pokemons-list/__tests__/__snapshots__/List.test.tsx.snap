// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<PokemonsList /> Renders 1`] = `
<div
  class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 css-172fket-MuiPaper-root"
  data-testid="pokemons-list-paper"
>
  <div
    class="MuiTableContainer-root css-rorn0c-MuiTableContainer-root"
  >
    <table
      class="MuiTable-root MuiTable-stickyHeader css-xn82ks-MuiTable-root"
    >
      <thead
        class="MuiTableHead-root css-15wwp11-MuiTableHead-root"
      >
        <tr
          class="MuiTableRow-root MuiTableRow-head css-1q1u3t4-MuiTableRow-root"
        >
          <th
            class="MuiTableCell-root MuiTableCell-head MuiTableCell-stickyHeader MuiTableCell-sizeMedium css-u2m7gh-MuiTableCell-root"
            data-testid="pokemons-list-head-cell"
            scope="col"
          >
            <div
              class="css-gmwslw-MuiStack-root"
            >
              Name
              <span
                class="MuiCircularProgress-root MuiCircularProgress-indeterminate MuiCircularProgress-colorPrimary css-18lrjg1-MuiCircularProgress-root"
                role="progressbar"
                style="width: 20px; height: 20px;"
              >
                <svg
                  class="MuiCircularProgress-svg css-1idz92c-MuiCircularProgress-svg"
                  viewBox="22 22 44 44"
                >
                  <circle
                    class="MuiCircularProgress-circle MuiCircularProgress-circleIndeterminate css-176wh8e-MuiCircularProgress-circle"
                    cx="44"
                    cy="44"
                    fill="none"
                    r="20.2"
                    stroke-width="3.6"
                  />
                </svg>
              </span>
            </div>
          </th>
        </tr>
      </thead>
      <tbody
        class="MuiTableBody-root css-apqrd9-MuiTableBody-root"
      />
    </table>
  </div>
  <div
    class="MuiTablePagination-root css-jtlhu6-MuiTablePagination-root"
    data-testid="pokemons-list-pagination"
  >
    <div
      class="MuiToolbar-root MuiToolbar-gutters MuiToolbar-regular MuiTablePagination-toolbar css-11bfvty-MuiToolbar-root-MuiTablePagination-toolbar"
    >
      <div
        class="MuiTablePagination-spacer css-1psng7p-MuiTablePagination-spacer"
      />
      <p
        class="MuiTablePagination-selectLabel css-pdct74-MuiTablePagination-selectLabel"
        id=":r1:"
      >
        Rows per page:
      </p>
      <div
        class="MuiInputBase-root MuiInputBase-colorPrimary css-16c50h-MuiInputBase-root-MuiTablePagination-select"
        variant="standard"
      >
        <div
          aria-expanded="false"
          aria-haspopup="listbox"
          aria-labelledby=":r1: :r0:"
          class="MuiTablePagination-select MuiSelect-select MuiSelect-standard MuiInputBase-input css-194a1fa-MuiSelect-select-MuiInputBase-input"
          id=":r0:"
          role="button"
          tabindex="0"
        >
          10
        </div>
        <input
          aria-hidden="true"
          class="MuiSelect-nativeInput css-yf8vq0-MuiSelect-nativeInput"
          tabindex="-1"
          value="10"
        />
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiTablePagination-selectIcon MuiSelect-icon MuiSelect-iconStandard css-pqjvzy-MuiSvgIcon-root-MuiSelect-icon"
          data-testid="ArrowDropDownIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M7 10l5 5 5-5z"
          />
        </svg>
      </div>
      <p
        class="MuiTablePagination-displayedRows css-levciy-MuiTablePagination-displayedRows"
      >
        0â€“0 of 0
      </p>
      <div
        class="MuiTablePagination-actions"
      >
        <button
          aria-label="Go to previous page"
          class="MuiButtonBase-root Mui-disabled MuiIconButton-root Mui-disabled MuiIconButton-colorInherit MuiIconButton-sizeMedium css-zylse7-MuiButtonBase-root-MuiIconButton-root"
          disabled=""
          tabindex="-1"
          title="Go to previous page"
          type="button"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
            data-testid="KeyboardArrowLeftIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M15.41 16.09l-4.58-4.59 4.58-4.59L14 5.5l-6 6 6 6z"
            />
          </svg>
        </button>
        <button
          aria-label="Go to next page"
          class="MuiButtonBase-root Mui-disabled MuiIconButton-root Mui-disabled MuiIconButton-colorInherit MuiIconButton-sizeMedium css-zylse7-MuiButtonBase-root-MuiIconButton-root"
          disabled=""
          tabindex="-1"
          title="Go to next page"
          type="button"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
            data-testid="KeyboardArrowRightIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M8.59 16.34l4.58-4.59-4.58-4.59L10 5.75l6 6-6 6z"
            />
          </svg>
        </button>
      </div>
    </div>
  </div>
</div>
`;
